framework/contrib/hit/parse.cc

```diff
@@ -997,6 +999,188 @@ explode(Node * n)
+
+void
+Formatter::walkPatternConfig(const std::string & prefix, Node * n)
+{
+  std::vector<std::string> order;
+  for (auto child : n->children())
+  {
+    order.push_back(child->path());
+    if (child->type() == NodeType::Section)
+    {
+      auto subpath = prefix + "/" + child->path();
+      if (prefix == "")
+        subpath = child->path();
+      walkPatternConfig(subpath, child);
+    }
+  }
+
+  addPattern(prefix, order);
+}
+
+Formatter::Formatter(const std::string & fname, const std::string & hit_config)
+  : canonical_section_markers(true), line_length(100), indent_string("  ")
+{
+  std::unique_ptr<hit::Node> root(hit::parse(fname, hit_config));
+  if (root->find("format/indent_string"))
+    indent_string = root->param<std::string>("format/indent_string");
+  if (root->find("format/line_length"))
+    line_length = root->param<int>("format/line_length");
+  if (root->find("format/canonical_section_markers"))
+    canonical_section_markers = root->param<bool>("format/canonical_section_markers");
+  if (root->find("format/sorting"))
+    walkPatternConfig("", root->find("format/sorting"));
+}
+
+std::string
+Formatter::format(const std::string & fname, const std::string & input)
+{
+  std::unique_ptr<hit::Node> root(hit::parse(fname, input));
+
+  TokenClearer tc;
+  if (canonical_section_markers)
+    root->walk(&tc, hit::NodeType::Section);
+
+  root->walk(this, hit::NodeType::All);
+  return root->render(0, indent_string, line_length);
+}
+
+void
+Formatter::addPattern(const std::string & section, const std::vector<std::string> & order)
+{
+  _patterns.push_back({section, order});
+}
+
+void
+Formatter::walk(const std::string & fullpath, const std::string & /*nodepath*/, Node * n)
+{
+  for (auto & pattern : _patterns)
+  {
+    if (!matches(fullpath, pattern.regex, true))
+      continue;
+
+    std::vector<std::string> frontorder;
+    std::vector<std::string> backorder;
+    bool onfront = true;
+    for (auto & field : pattern.order)
+    {
+      if (field == "**")
+      {
+        onfront = false;
+        continue;
+      }
+      else if (onfront)
+        frontorder.push_back(field);
+      else
+        backorder.push_back(field);
+    }
+
+    auto nodes = n->children();
+    std::vector<Node *> fronthalf;
+    std::vector<Node *> unused;
+    sortGroup(nodes, frontorder, fronthalf, unused);
+
+    std::vector<Node *> backhalf;
+    nodes = unused;
+    unused.clear();
+    sortGroup(nodes, backorder, backhalf, unused);
+
+    std::vector<Node *> children;
+    children.insert(children.end(), fronthalf.begin(), fronthalf.end());
+    children.insert(children.end(), unused.begin(), unused.end());
+    children.insert(children.end(), backhalf.rbegin(), backhalf.rend());
+
+    for (unsigned int i = 0; i < children.size(); i++)
+      children[i] = children[i]->clone();
+
+    for (auto child : n->children())
+      delete child;
```

Comment:
What is this business?  Use `std::unique_ptr`

Reply:
You don't need to free it.  The tree manages its own memory (manually).  But here the `delete` operation is actually removing nodes from the tree.

Reply:
We crossed comments there.

In modern times there is simply no reason for this pattern (to call a function and then free stuff).  If it would require a rewrite of HIT then that's what it requires (I doubt that though... I'm sure you can retrofit your current design with some smart pointers... even if they're `std::shared_ptr`).

If that's outside the scope of this PR then I can accept that - but I'd like to see an Issue created by you for removal of this pattern from HIT first.

Reply:
OK - so we have a plan forward here:

I talked to Robert and he says that the internal tree structure manages it's own memory completely. Deleting a tree will delete all memory in the tree. What we'd like to do, is get to the point where you don't have to even delete the tree... This is important because with tree manipulation, any node in the tree has the potential to become a root which requires a delete:

1) Handing back memory that needs to be deleted is a no-go in 2018. John, David and I (as well as you) are all in agreement about this. A good compromise to rewriting all of his is to have any manipulation that "creates" or orphans a subtree (e.g. clone) will return back a smart pointer to that subtree. The internal nodes which are managed by their parent can remain dump pointers (I'll leave it up to Robert to decide which flavor). 

2) Any inconsistencies or other issues that arise from using a smart pointer to the root will be addressed. The end goal is external users should not have to alloc or delete memory. Instead they should create or remove nodes.

3) Move away from using the memory keyword "delete" as a tree manipulation function. Instead, let's make the Node API a little smarter and use a method instead.

If I've stated any of these bullets incorrectly, please correct them. Let's make sure we get the right compromise design moving forward.

At this time, rewriting the internal structure of the HIT tree to use all smart pointers is probably not a very good use of Robert's time. I do have other things I'd like him to focus on. I still think we should revisit this, perhaps in the future when we plan to move HIT outside of MOOSE or up into libMesh. that'll be a good time for an internal representation rewrite as we'll have probably learned quite a bit.

I think we may have an issue open for the internal representation rewrite. @rwcarlsen will you find that ticket and make any relevant design comments on that ticket? Let's not hold up this PR for the time being.

Reply:
👍